src/pykeen/models/unimodal/rgcn.py

```diff
@@ -105,7 +105,442 @@ def symmetric_edge_weights(source: torch.LongTensor, target: torch.LongTensor) -
+
+    def __init__(
+        self,
+        input_dim: int,
+        num_relations: int,
+        num_blocks: Optional[int] = None,
+        output_dim: Optional[int] = None,
+    ):
+        """Initialize the layer.
+
+        :param input_dim: >0
+            The input dimension.
+        :param num_relations: >0
+            The number of relations.
+        :param num_blocks: >0
+            The number of blocks to use. Has to be a divisor of input_dim.
+        :param output_dim: >0
+            The output dimension. If None is given, defaults to input_dim.
+        """
+        super().__init__(
+            input_dim=input_dim,
+            num_relations=num_relations,
+            output_dim=output_dim,
+        )
+
+        if num_blocks is None:
+            logging.info('Using a heuristic to determine the number of blocks.')
+            num_blocks = min(i for i in range(2, input_dim + 1) if input_dim % i == 0)
+
+        block_size, remainder = divmod(input_dim, num_blocks)
+        if remainder != 0:
+            raise NotImplementedError(
+                'With block decomposition, the embedding dimension has to be divisible by the number of'
+                f' blocks, but {input_dim} % {num_blocks} != 0.',
+            )
+
+        self.blocks = nn.Parameter(
+            data=torch.empty(
+                num_relations + 1,
+                num_blocks,
+                block_size,
+                block_size,
+            ), requires_grad=True)
+        self.num_blocks = num_blocks
+        self.block_size = block_size
+
+    def reset_parameters(self):  # noqa: D102
+        block_size = self.blocks.shape[-1]
+        # Xavier Glorot initialization of each block
+        std = torch.sqrt(torch.as_tensor(2.)) / (2 * block_size)
+        nn.init.normal_(self.blocks, std=std)
+
+    def forward(
+        self,
+        x: torch.FloatTensor,
+        node_keep_mask: Optional[torch.BoolTensor],
+        source: torch.LongTensor,
+        target: torch.LongTensor,
+        edge_type: torch.LongTensor,
+        edge_weights: Optional[torch.FloatTensor] = None,
+    ) -> torch.FloatTensor:  # noqa: D102
+        # view as blocks
+        x = x.view(-1, self.num_blocks, self.block_size)
+
+        # self-loop first
+        out = torch.zeros_like(x)
+        w = self.blocks[-1]
+        if node_keep_mask is not None:
+            out[node_keep_mask] = torch.einsum('nbi,bij->nbj', x[node_keep_mask], w)
+        else:
+            out = torch.einsum('nbi,bij->nbj', x, w)
+
+        # other relations
+        for r in range(self.num_relations):
+            specific = _reduce_relation_specific(
+                relation=r,
+                source=source,
+                target=target,
+                edge_type=edge_type,
+                edge_weights=edge_weights,
+            )
+
+            # skip relations without edges
+            if specific is None:
```

Comment:
I think there's a typo here. shouldn't you directly unpack `specific` then check `source_r is None`? It looks like that in another place

Reply:
You are right. 
Or return just a single `None` in case of `source_r=None`.

Reply:
I like returning a tuple with `None, None, None` because it allows the unpacking of the results to be safe

Reply:
Addressed in 208c53ab